language/src/typechecker.rs

```diff
@@ -2220,8 +2382,109 @@ fn typecheck_question_mark(
+                        return Err(());
+                    }
+                }
+            }
+            (
+                (Borrowing::Consumed, _),
+                (BaseTyp::Named((TopLevelIdent { string: name, .. }, _), Some(args)), _),
+            ) if name == "Result" && args.len() == 2 => {
+                let ok_typ = &args[0];
+                let err_typ = &args[1];
+                match return_typ {
+                    (
+                        BaseTyp::Named(
+                            (
+                                TopLevelIdent {
+                                    string: return_name,
+                                    ..
+                                },
+                                _,
+                            ),
+                            Some(return_args),
+                        ),
+                        _,
+                    ) if return_name == "Result" && return_args.len() == 2 => {
+                        let err_typ_ret = &args[1];
+
+                        match unify_types(
+                            sess,
+                            &((Borrowing::Consumed, err_typ.1.clone()), err_typ.clone()),
+                            &(
+                                (Borrowing::Consumed, err_typ_ret.1.clone()),
+                                err_typ_ret.clone(),
+                            ),
+                            &HashMap::new(),
+                            top_level_context,
+                        )? {
+                            Some(_) => Ok(CarrierTyp::Result(ok_typ.clone(), err_typ_ret.clone())),
+                            None => {
+                                sess.span_rustspec_err(
+                                    expr_span,
+                                    format!(
+                                        "the type returned in case of error by this \
+                                        expression is {}, expected {}",
+                                        err_typ.0, err_typ_ret.0,
+                                    )
+                                    .as_str(),
+                                );
+                                return Err(());
+                            }
+                        }
+                    }
+                    _ => {
+                        sess.span_rustspec_err(
+                            return_typ.1,
+                            format!(
+                                "expected a result type for this \
+                    return type because of a question mark in the function, got {}",
+                                return_typ.0,
+                            )
+                            .as_str(),
+                        );
+                        return Err(());
+                    }
+                }
+            }
+            (
+                (Borrowing::Consumed, _),
+                (BaseTyp::Named((TopLevelIdent { string: name, .. }, _), expr_typ), _),
+            ) => Err(()),
```

Comment:
I changed this to do unification, I had an issue with AES not working without this.